dnstable/query.c

```diff
@@ -528,6 +529,123 @@ increment_key(ubuf *key, size_t pos)
 	return (dnstable_res_failure);
 }
 
+#define SL_MATCHES	0		/* Label matches. */
+#define SL_NOMATCH	1		/* Not a match. */
+#define SL_FAILURE	2		/* Error detected. */
+
+/*
+ * Used with a "single label" wildcard search to determine the action to take
+ * for a given matching key.
+ * Either:
+ *  + Matches  -- One and only one extra label.
+ *  + No match -- Restart search.
+ *  + Failure  -- Some error.
+ *
+ * Return-value indicates the action to take.
+ */
+static int
+key_label_check(struct query_iter *it, const uint8_t *key, size_t len_key, bool seek_iter2)
+{
+	size_t qks, next_len;
+
+	/*
+	 * The "key" in the query is label-reversed, and the wildcard char is removed.
+	 * e.g. For rrset "+.example.com", the key will be:
+	 *   \x00\x03com\x07example
+	 *
+	 * The key that has been found (and is being checked here) will start with these
+	 * same bytes, and should have at the very least a trailing length-byte as well.
+	 */
+	qks = ubuf_size(it->key);		/* Byte-size of original key. */
+
+	if (qks >= len_key)
+		return (SL_NOMATCH);
+
+	/*
+	 * No more labels on this key.
+	 *   \x00\x03com\x07example\x00
+	 */
+	if (key[qks] == '\0') {
+		ubuf *seek_key;
+
+		/*
+		 * Now skip all keys where the leading-part is the same as this key.
+		 * e.g. seek to key
+		 *   \x00\x03com\x07example\x01
+		 */
+
+		/* Create a new key. */
+		if (it->key2 == NULL)
+			it->key2 = ubuf_init(qks + 1);
+		seek_key = it->key2;
+		ubuf_clip(seek_key, 0);
+
+		ubuf_reserve(seek_key, qks + 1);
+		ubuf_append(seek_key, key, qks);
+		ubuf_add(seek_key, '\1');
+
+		/* Seek to the newly generated key. */
+		if (mtbl_iter_seek(seek_iter2 ? it->m_iter2 : it->m_iter,
+					ubuf_data(seek_key),
+					ubuf_size(seek_key)) != mtbl_res_success)
+			return (SL_FAILURE);
+
+		/* Restart processing starting from the new key. */
+		return (SL_NOMATCH);
+	}
+
+	/* From here, there is at least one more label. */
+
+	/* The position of the "length" byte after the first extra label. */
+	next_len = qks + key[qks] + 1;
+
+	/*
+	 * If the key is malformed, treat it as if there are no more labels.
```

Comment:
Note that this scenario is possible for well-formed rdata entries of non-name-based rrtypes, for example a 13-byte NULL entry like the following:

`\x03 \x03com\x07example\xFF \x0a \x03com\x07example\x00 \x00\x0d`

I believe returning without seek to the caller is appropriate here, because this cannot happen with well-formed rdata _name_ entries, and the caller will filter out the offending entry.